text/2978-stack_vec.md

```diff
@@ -0,0 +1,185 @@
+
+Just like `Vec`, `StackVec` is a primitive vector where high-level structures can use it as a building block. For example, a stack-based matrix or binary heap.
+
+### Useful in the real world
+
+`arrayvec` is one of the most downloaded project of `crates.io` and is used by thousand of projects, including Rustc itself.
+
+### Compiler internals
+
+Unstable features can be used internally to make operations more efficient, e.g., `specialization`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Instead of relying on a heap-allocator, stack-based memory area is added and removed on-demand in a last-in-first-out (LIFO) order according to the calling workflow of a program. Let's illustrate an imaginary function `super()` that allocates 32 bits before calling another function named `sub()` that also allocates 32 bits of auxiliary memory:
+
+```txt
+0 bits           32 bits          64 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+                 | super |        |  sub  |
+                 ---------        ---------
+                                  | super |
+                                  ---------
+```
+
+* Now `super()` returns and the exactly same thing happens in reverse order, dropping everything when out of scope.
+
+```txt
+64 bits          32 bits          0 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+|  sub  |        | super |
+---------        ---------
+| super |
+---------
+```
+
+`StackVec` takes advantage of this predictable behavior to reserve an exactly amount of uninitialized bytes up-front and these bytes form a buffer where elements can be included dynamically.
+
+```rust
+fn main() {
+    // `stack_vec` has a pre-allocated memory of 2048 bits that can store up to 64 decimals.
+    let mut stack_vec: StackVec<i32, 64> = StackVec::new();
+
+    // Although reserved, there isn't anything explicitly stored yet
+    assert_eq!(stack_vec.len(), 0);
+
+    // Initializes the first 32 bits with a simple '1' decimal or
+    // 00000000 00000000 00000000 00000001 bits
+    stack_vec.push(1);
+
+    // Our vector memory is now split into a 32/2016 pair of initialized and
+    // uninitialized memory respectively
+    assert_eq!(stack_vec.len(), 1);
+}
+```
+
+Of course, fixed buffers lead to some inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The most natural module for `StackVec` is `core::collections` with an API that basically mimics most of the current `Vec<T>` surface.
+
+```rust
+// core::collections
+
+pub struct StackVec<T, const N: usize> {
+    data: MaybeUninit<[T, N]>,
+    len: usize
+}
+
+impl<T, const N: usize> StackVec<T, N> {
+    // Much of the `Vec` API goes here
```

Comment:
Mirroring the `Vec` API doesn't really work that well for a structure with bounded capacity, since `push` can fail. What should happen then?

Reply:
My reasoning is: There must be two methods for everything that tries to add elements. One method panics at run-time and other one returns a `Result`.

```rust
fn push(&mut self, element: T) {  ... } // Panics at run-time
fn try_push(&mut self, element: T) -> Result<(), Error> { ... } // Ok if successful 
```

Reply:
> Mirroring the `Vec` API doesn't really work that well for a structure with bounded capacity

That's not really true at all, frankly. Many of the functions can in fact be exactly the same in every way, and for the ones where regular `Vec` would have simply increased in capacity upon becoming full, you have the option of either simply panicking or returning something like `Err` or `None`. Or doing both at different times, if it's the case that you simultaneously provide `push` and `try_push` or `insert` and `try_insert`, for example.